# Data Cleaning Practice 3

In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
data = pd.read_csv('messy_data2.csv')

In [4]:
df = data.copy(deep=True)

# 

# Data Inspection

In [5]:
df.head()

,Employee_ID,Name,Age,Salary,Department,City,Email,Phone_Number,Join_Date,Experience_Years,Performance_Score,Gender
0,EMP1000,Yasmin Thakur,NaN,NaN,Support,Bangalore,wthaker@yahoo.com,NaN,2022-01-30,0.4,10.0,NaN
1,NaN,Stuvan,150.0,unknown,Marketing,Delhi,umangbalakrishnan@viswanathan.biz,1.234500e+04,NaN,NaN,2.0,Other
2,EMP1002,NaN,-5.0,-10000,NaN,Bangalore,NaN,4.470941e+09,32-13-2023,NaN,NaN,NaN
3,NaN,Yuvaan Raju,NaN,90861,NaN,Mumbai,venkateshpihu@dash.com,NaN,32-13-2023,NaN,NaN,Other
4,NaN,NaN,59.0,NaN,HR,Pune,NaN,1.000000e+15,NaN,10.3,3.0,Female


In [6]:
df.tail()

,Employee_ID,Name,Age,Salary,Department,City,Email,Phone_Number,Join_Date,Experience_Years,Performance_Score,Gender
995,EMP1067,NaN,-5.0,unknown,Finance,NaN,invalid_email,1.234500e+04,32-13-2023,-3.0,1.0,Male
996,EMP1996,Pari,150.0,NaN,HR,Bangalore,invalid_email,1.903628e+09,2024-06-05,NaN,10.0,Male
997,EMP1099,NaN,NaN,NaN,NaN,Bangalore,pranay@gmail,1.234500e+04,32-13-2023,7.7,10.0,Male
998,EMP1998,Nishith,150.0,NaN,HR,NaN,kartik@gmail,1.000000e+15,32-13-2023,3.7,10.0,Female
999,NaN,Gokul Kota,150.0,50626,Finance,Pune,NaN,NaN,NaN,NaN,NaN,NaN


In [7]:
 df.sample(10)

,Employee_ID,Name,Age,Salary,Department,City,Email,Phone_Number,Join_Date,Experience_Years,Performance_Score,Gender
666,EMP1053,NaN,NaN,-10000,Support,Hyderabad,NaN,9.128980e+11,NaN,-3.0,2.0,NaN
494,EMP1494,Parinaaz,150.0,-10000,IT,Bangalore,NaN,NaN,32-13-2023,-3.0,NaN,Male
311,NaN,NaN,22.0,-10000,HR,Delhi,manikya@gmail,1.000000e+15,NaN,-3.0,NaN,NaN
211,EMP1067,Akarsh,28.0,NaN,Finance,Hyderabad,duayakshit@gmail.com,1.234500e+04,NaN,NaN,3.0,Female
189,EMP1189,Taran,-5.0,NaN,Sales,Hyderabad,NaN,1.234500e+04,2021-09-16,13.4,10.0,NaN
278,NaN,NaN,NaN,-10000,Sales,Mumbai,vivaanrana@lanka.info,1.000000e+15,2021-09-22,-3.0,NaN,Other
448,EMP1448,Aarav,150.0,-10000,NaN,NaN,NaN,1.000000e+15,32-13-2023,NaN,10.0,Other
554,EMP1026,NaN,53.0,NaN,Marketing,NaN,invalid_email,1.000000e+15,NaN,-3.0,NaN,Female
292,NaN,Diya Dhar,54.0,unknown,Support,Mumbai,nirvi@gmail,NaN,32-13-2023,NaN,NaN,Other
681,EMP1681,Jayesh,-5.0,NaN,IT,Delhi,invalid_email,1.234500e+04,2022-01-06,10.9,NaN,NaN


# 

In [8]:
df.dtypes

Employee_ID           object
Name                  object
Age                  float64
Salary                object
Department            object
City                  object
Email                 object
Phone_Number         float64
Join_Date             object
Experience_Years     float64
Performance_Score    float64
Gender                object
dtype: object

In [9]:
df.describe()

,Age,Phone_Number,Experience_Years,Performance_Score
count,765.000000,7.570000e+02,647.000000,652.000000
mean,67.067974,3.409361e+14,1.887326,6.450920
std,66.675098,4.742139e+14,6.083940,3.653467
min,-5.000000,1.234500e+04,-3.000000,1.000000
25%,-5.000000,1.234500e+04,-3.000000,3.000000
50%,44.000000,8.513171e+09,-3.000000,5.000000
75%,150.000000,1.000000e+15,7.500000,10.000000
max,150.000000,1.000000e+15,14.900000,10.000000


In [10]:
df.isnull().sum()

Employee_ID          329
Name                 499
Age                  235
Salary               264
Department           143
City                 143
Email                283
Phone_Number         243
Join_Date            321
Experience_Years     353
Performance_Score    348
Gender               417
dtype: int64

In [11]:
df.isnull().mean()*100

Employee_ID          32.9
Name                 49.9
Age                  23.5
Salary               26.4
Department           14.3
City                 14.3
Email                28.3
Phone_Number         24.3
Join_Date            32.1
Experience_Years     35.3
Performance_Score    34.8
Gender               41.7
dtype: float64

In [12]:
df.duplicated().sum()

np.int64(0)

In [14]:
df['Employee_ID'].duplicated().sum()

np.int64(584)

In [15]:
df[df['Employee_ID'].duplicated()]

,Employee_ID,Name,Age,Salary,Department,City,Email,Phone_Number,Join_Date,Experience_Years,Performance_Score,Gender
3,NaN,Yuvaan Raju,NaN,90861,NaN,Mumbai,venkateshpihu@dash.com,NaN,32-13-2023,NaN,NaN,Other
4,NaN,NaN,59.0,NaN,HR,Pune,NaN,1.000000e+15,NaN,10.3,3.0,Female
8,NaN,NaN,NaN,148085,NaN,Chennai,kismat@gmail,1.234500e+04,2025-09-14,5.6,10.0,Male
12,NaN,NaN,NaN,NaN,NaN,Pune,NaN,1.234500e+04,2019-07-09,NaN,NaN,NaN
13,NaN,Lakshit Bajwa,59.0,unknown,Support,Delhi,charvi95@boase.biz,1.234500e+04,2018-06-07,NaN,NaN,Female
...,...,...,...,...,...,...,...,...,...,...,...,...
993,EMP1046,Anvi Bhavsar,150.0,unknown,IT,Pune,umang@gmail,1.000000e+15,2018-05-04,NaN,4.0,Female
994,NaN,Charvi,150.0,110570,NaN,Mumbai,qkumer@ramaswamy.com,1.234500e+04,2018-10-31,-3.0,NaN,Male
995,EMP1067,NaN,-5.0,unknown,Finance,NaN,invalid_email,1.234500e+04,32-13-2023,-3.0,1.0,Male
997,EMP1099,NaN,NaN,NaN,NaN,Bangalore,pranay@gmail,1.234500e+04,32-13-2023,7.7,10.0,Male


In [16]:
df[df['Employee_ID'] == 'EMP1099']

,Employee_ID,Name,Age,Salary,Department,City,Email,Phone_Number,Join_Date,Experience_Years,Performance_Score,Gender
247,EMP1099,Anahi,150.0,NaN,IT,Hyderabad,NaN,1.234500e+04,NaN,NaN,2.0,Male
312,EMP1099,NaN,NaN,63204,HR,Chennai,invalid_email,1.000000e+15,NaN,10.5,2.0,NaN
902,EMP1099,NaN,56.0,118669,Support,NaN,NaN,1.234500e+04,NaN,10.3,1.0,Female
997,EMP1099,NaN,NaN,NaN,NaN,Bangalore,pranay@gmail,1.234500e+04,32-13-2023,7.7,10.0,Male


In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Employee_ID        671 non-null    object 
 1   Name               501 non-null    object 
 2   Age                765 non-null    float64
 3   Salary             736 non-null    object 
 4   Department         857 non-null    object 
 5   City               857 non-null    object 
 6   Email              717 non-null    object 
 7   Phone_Number       757 non-null    float64
 8   Join_Date          679 non-null    object 
 9   Experience_Years   647 non-null    float64
 10  Performance_Score  652 non-null    float64
 11  Gender             583 non-null    object 
dtypes: float64(4), object(8)
memory usage: 93.9+ KB


# 

# 

## Data Cleaning

In [18]:
df.columns

Index(['Employee_ID', 'Name', 'Age', 'Salary', 'Department', 'City', 'Email',
       'Phone_Number', 'Join_Date', 'Experience_Years', 'Performance_Score',
       'Gender'],
      dtype='object')

In [19]:
df['Employee_ID'].head()

0    EMP1000
1        NaN
2    EMP1002
3        NaN
4        NaN
Name: Employee_ID, dtype: object

In [22]:
df['Employee_ID'] = df['Employee_ID'].str.strip().str.upper()

In [24]:
df.drop_duplicates(subset='Employee_ID', inplace=True)

In [26]:
df['Employee_ID'].duplicated().sum()

np.int64(0)

In [28]:
df['Employee_ID'].isnull().mean()

np.float64(0.002403846153846154)

In [29]:
df['Employee_ID'].isnull().sum()

np.int64(1)

In [32]:
df['Employee_ID'].head(10)

0     EMP1000
1         NaN
2     EMP1002
5     EMP1004
6     EMP1068
7     EMP1007
9     EMP1076
10    EMP1084
11    EMP1008
14    EMP1035
Name: Employee_ID, dtype: object

In [33]:
df.dropna(subset='Employee_ID', inplace=True)

In [34]:
df['Employee_ID'].isnull().sum()

np.int64(0)

# 

In [35]:
df.columns

Index(['Employee_ID', 'Name', 'Age', 'Salary', 'Department', 'City', 'Email',
       'Phone_Number', 'Join_Date', 'Experience_Years', 'Performance_Score',
       'Gender'],
      dtype='object')

In [37]:
df['Age'].unique()

array([ nan,  -5.,  35., 150.,  19.,  57.,  47.,  54.,  32.,  43.,  53.,
        44.,  38.,  24.,  42.,  21.,  34.,  27.,  41.,  59.,  28.,  56.,
        39.,  48.,  60.,  20.,  46.,  26.,  25.,  45.,  37.,  55.,  50.,
        18.,  40.,  31.,  58.,  52.,  22.,  29.,  36.,  51.,  23.])

In [38]:
df['Age'].dtypes

dtype('float64')

In [39]:
df['Age'] = df['Age'].astype('Int64')

In [40]:
df['Age'].dtypes

Int64Dtype()

In [41]:
df['Age'].unique()

<IntegerArray>
[<NA>,   -5,   35,  150,   19,   57,   47,   54,   32,   43,   53,   44,   38,
   24,   42,   21,   34,   27,   41,   59,   28,   56,   39,   48,   60,   20,
   46,   26,   25,   45,   37,   55,   50,   18,   40,   31,   58,   52,   22,
   29,   36,   51,   23]
Length: 43, dtype: Int64

In [42]:
df.loc[df['Age']< 0,'Age'] = np.nan

In [43]:
df.loc[df['Age']>110,'Age'] = np.nan

In [44]:
df['Age'].unique()

<IntegerArray>
[<NA>,   35,   19,   57,   47,   54,   32,   43,   53,   44,   38,   24,   42,
   21,   34,   27,   41,   59,   28,   56,   39,   48,   60,   20,   46,   26,
   25,   45,   37,   55,   50,   18,   40,   31,   58,   52,   22,   29,   36,
   51,   23]
Length: 41, dtype: Int64

# 

In [45]:
df.columns

Index(['Employee_ID', 'Name', 'Age', 'Salary', 'Department', 'City', 'Email',
       'Phone_Number', 'Join_Date', 'Experience_Years', 'Performance_Score',
       'Gender'],
      dtype='object')

In [48]:
df['Salary'].dtypes.name

'object'

In [49]:
df['Salary'] = pd.to_numeric(df['Salary'],errors='coerce')

In [54]:
df[df['Salary']< 0].size

1332

In [56]:
df.loc[df['Salary']< 0,'Salary'] = np.nan

In [58]:
df['Salary'].isnull().mean()*100

np.float64(77.59036144578313)

# 

In [59]:
df['Department'].unique()

array(['Support', nan, 'HR', 'IT', 'Finance', 'Marketing', 'Sales'],
      dtype=object)

In [60]:
df['Department'] = df['Department'].str.title()

In [61]:
df['Department'] = df['Department'].fillna(df['Department'].mode()[0])

In [62]:
df['Department'].unique()

array(['Support', 'Hr', 'It', 'Finance', 'Marketing', 'Sales'],
      dtype=object)

In [63]:
df['Department'].isnull().sum()

np.int64(0)

# 

In [64]:
df['City'].unique()

array(['Bangalore', 'Chennai', 'Hyderabad', 'Delhi', 'Pune', 'Mumbai',
       nan], dtype=object)

In [65]:
df['City'] = df['City'].str.strip().str.title()

In [66]:
df['City'] = df.groupby('Department')['City'].transform(lambda x: x.fillna(x.mode()[0]))

In [67]:
df['City'].unique()

array(['Bangalore', 'Chennai', 'Hyderabad', 'Delhi', 'Pune', 'Mumbai'],
      dtype=object)

In [68]:
df['City'].isnull().sum()

np.int64(0)

# 

In [71]:
df['Email'].isnull().sum()

np.int64(114)

In [72]:
df.duplicated().sum()

np.int64(0)

In [73]:
df = df.drop_duplicates(subset='Email', keep='first')

# 

In [75]:
df['Experience_Years']

0       0.4
2       NaN
7       NaN
9      11.0
11      NaN
       ... 
962    -3.0
976    -3.0
986    -3.0
989     NaN
998     3.7
Name: Experience_Years, Length: 177, dtype: float64

In [76]:
df['Experience_Years'] = np.abs(df['Experience_Years'])

C:\Users\Omkar Yallurkar\AppData\Local\Temp\ipykernel_6836\3548653902.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Experience_Years'] = np.abs(df['Experience_Years'])


In [79]:
df['Experience_Years'].isnull().sum()

np.int64(64)

In [83]:
df['Experience_Years'] = (df.groupby(['Department', 'City'])['Experience_Years'].transform(lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else x.median())))

In [84]:
df['Experience_Years'].isnull().sum()

np.int64(3)

In [86]:
df['Experience_Years'] = df['Experience_Years'].fillna(df['Experience_Years'].mode()[0])

In [87]:
df['Experience_Years'].isnull().sum()

np.int64(0)

# 

In [88]:
df['Phone_Number'].dtype

dtype('float64')

In [90]:
df['Phone_Number'].head(10)

0              NaN
2     4.470941e+09
7     1.000000e+15
9     9.811370e+09
11    1.234500e+04
14    5.642615e+09
17    1.234500e+04
20    5.143210e+09
23             NaN
24    1.000000e+15
Name: Phone_Number, dtype: float64

In [91]:
df['Phone_Number'].tail()

962    1.000000e+15
976    1.234500e+04
986             NaN
989             NaN
998    1.000000e+15
Name: Phone_Number, dtype: float64

In [92]:
df = df.drop(columns='Phone_Number')

In [93]:
df.columns

Index(['Employee_ID', 'Name', 'Age', 'Salary', 'Department', 'City', 'Email',
       'Join_Date', 'Experience_Years', 'Performance_Score', 'Gender'],
      dtype='object')

# 

In [94]:
df['Join_Date'].unique()

array(['2022-01-30', '32-13-2023', nan, '2021-08-26', '2019-07-19',
       '2017-09-27', '2023-07-13', '2020-05-26', '2020-03-10',
       '2024-07-30', '2026-04-01', '2018-11-17', '2024-08-21',
       '2023-07-31', '2025-02-14', '2019-02-18', '2022-03-18',
       '2016-06-20', '2021-03-19', '2020-08-03', '2022-06-17',
       '2021-05-30', '2023-06-17', '2021-05-07', '2023-01-02',
       '2023-03-12', '2023-08-21', '2022-02-08', '2020-08-29',
       '2016-12-02', '2026-03-07', '2021-11-25', '2017-03-02',
       '2025-03-03', '2022-07-25', '2020-01-12', '2023-10-17',
       '2020-12-18', '2022-11-03', '2024-01-12', '2017-08-17',
       '2020-09-04', '2018-05-02', '2024-09-26', '2023-08-03',
       '2019-05-12', '2021-03-31', '2026-04-07', '2022-11-25',
       '2017-08-15', '2021-12-26', '2020-06-07', '2021-09-27',
       '2021-05-18', '2017-08-31', '2024-03-30', '2019-05-14'],
      dtype=object)

In [95]:
df['Join_Date'].dtypes

dtype('O')

In [96]:
df['Join_Date'] = pd.to_datetime(df['Join_Date'],errors='coerce')

In [98]:
df['Join_Date'].dtypes.name

'datetime64[ns]'

In [99]:
df['Join_Date'].isnull().sum()

np.int64(122)

In [100]:
df['Join_Date'].isnull().mean()*100

np.float64(68.92655367231639)

In [102]:
df['Join_Date'] = (df.groupby(['Department', 'City'])['Join_Date'].transform(lambda x: x.fillna(x.median())))

C:\Users\Omkar Yallurkar\anaconda3\Lib\site-packages\numpy\lib\_nanfunctions_impl.py:1241: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
C:\Users\Omkar Yallurkar\anaconda3\Lib\site-packages\numpy\lib\_nanfunctions_impl.py:1241: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
C:\Users\Omkar Yallurkar\anaconda3\Lib\site-packages\numpy\lib\_nanfunctions_impl.py:1241: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
C:\Users\Omkar Yallurkar\anaconda3\Lib\site-packages\numpy\lib\_nanfunctions_impl.py:1241: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
C:\Users\Omkar Yallurkar\anaconda3\Lib\site-packages\numpy\lib\_nanfunctions_impl.py:1241: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
C:\Users\Omkar Yallurkar\anaconda3\Lib\site-packages\numpy\lib\_nanfunctions_imp

In [104]:
df['Performance_Score'].isnull().sum()

np.int64(65)

In [107]:
df[ 'Performance_Score'].unique()

array([10., nan,  4.,  5.,  1.,  2.,  3.])

In [108]:
df[ 'Performance_Score'] = df[ 'Performance_Score'].fillna(df[ 'Performance_Score'].mean())

In [109]:
df['Performance_Score'].isnull().sum()

np.int64(0)

# 

In [110]:
df['Gender'].unique()

array([nan, 'Female', 'Other', 'Male'], dtype=object)

In [111]:
df['Gender'].isna().sum()

np.int64(75)

In [ ]:
df['Gender'] = df.groupby(['Department','City'])['Gender'].transform(lambda a : )

In [112]:
df['Gender'] = (df.groupby(['Department', 'City'])['Gender'].transform(lambda x: x.fillna(
              x.mode().iloc[0] if not x.mode().empty else np.nan ) ) )

C:\Users\Omkar Yallurkar\AppData\Local\Temp\ipykernel_6836\1610265594.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Gender'] = (df.groupby(['Department', 'City'])['Gender'].transform(lambda x: x.fillna(


In [113]:
df['Gender'].unique()

array(['Female', 'Male', 'Other', nan], dtype=object)

In [114]:
df['Gender'].isna().sum()

np.int64(5)

In [115]:
df = df.dropna(subset='Gender')

In [116]:
df['Gender'].isna().sum()

np.int64(0)

# 

In [117]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 172 entries, 0 to 998
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Employee_ID        172 non-null    object        
 1   Name               87 non-null     object        
 2   Age                45 non-null     Int64         
 3   Salary             45 non-null     float64       
 4   Department         172 non-null    object        
 5   City               172 non-null    object        
 6   Email              171 non-null    object        
 7   Join_Date          154 non-null    datetime64[ns]
 8   Experience_Years   172 non-null    float64       
 9   Performance_Score  172 non-null    float64       
 10  Gender             172 non-null    object        
dtypes: Int64(1), datetime64[ns](1), float64(3), object(6)
memory usage: 16.3+ KB


In [118]:
df.isnull().sum()

Employee_ID            0
Name                  85
Age                  127
Salary               127
Department             0
City                   0
Email                  1
Join_Date             18
Experience_Years       0
Performance_Score      0
Gender                 0
dtype: int64

In [119]:
df['Age'] = (df.groupby('Department')['Age'].transform(lambda x: x.fillna(x.median())))

C:\Users\Omkar Yallurkar\AppData\Local\Temp\ipykernel_6836\1293470227.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Age'] = (df.groupby('Department')['Age'].transform(lambda x: x.fillna(x.median())))


In [120]:
df['Salary'] = (df.groupby('Department')['Salary'].transform(lambda x: x.fillna(x.median())))

C:\Users\Omkar Yallurkar\AppData\Local\Temp\ipykernel_6836\475869433.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Salary'] = (df.groupby('Department')['Salary'].transform(lambda x: x.fillna(x.median())))


In [121]:
df.isnull().sum()

Employee_ID           0
Name                 85
Age                   0
Salary                0
Department            0
City                  0
Email                 1
Join_Date            18
Experience_Years      0
Performance_Score     0
Gender                0
dtype: int64

In [122]:
df['Join_Date'] = (df.groupby('Department')['Join_Date'].transform(lambda x: x.fillna(x.median() if x.notna().sum() > 0 else pd.NaT)))

C:\Users\Omkar Yallurkar\AppData\Local\Temp\ipykernel_6836\277461103.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Join_Date'] = (df.groupby('Department')['Join_Date'].transform(lambda x: x.fillna(x.median() if x.notna().sum() > 0 else pd.NaT)))


In [123]:
df.isnull().sum()

Employee_ID           0
Name                 85
Age                   0
Salary                0
Department            0
City                  0
Email                 1
Join_Date             0
Experience_Years      0
Performance_Score     0
Gender                0
dtype: int64

In [124]:
df = df.dropna(subset='Email')

In [125]:
df.isnull().sum()

Employee_ID           0
Name                 84
Age                   0
Salary                0
Department            0
City                  0
Email                 0
Join_Date             0
Experience_Years      0
Performance_Score     0
Gender                0
dtype: int64

# 

# 

In [126]:
df.shape

(171, 11)

In [127]:
data.shape

(1000, 12)

##### The dataset was cleaned by handling missing values, invalid entries, duplicates, and datatype issues, but significant data loss occurred due to excessive missing and incorrect records, resulting in a final dataset shape of (171, 11).